OpenRA.Mods.Common/Traits/Attack/AttackBase.cs

```diff
@@ -80,6 +80,10 @@ protected virtual bool CanAttack(Actor self, Target target)
 			if (!HasAnyValidWeapons(target))
 				return false;
 
+			var layer = self.Location.Layer;
+			if (layer == CustomMovementLayerType.Tunnel || layer == CustomMovementLayerType.Subterranean)
```

Comment:
I just realized that i'm confusing matters here.  I'm saying that for unit A targeting unit B, unit A drops its target when unit B enters a tunnel.  These changes are all about what happens when unit A enters a tunnel, so ^ is a different issue.